File: app/services/ui/contrast.rb
module Ui
  # WCAG 2.1 contrast math. Used to pick the most-readable text color
  # against an arbitrary background, in both Ruby (for server-rendered
  # templates and for validating `Storefronts::Palette` at boot) and
  # JavaScript (for live-preview panels and any runtime color picker).
  #
  # The `INK_*` constants are Kitchef's two default text tokens —
  # `INK_LIGHT` is bone (used on dark backgrounds), `INK_DARK` is the
  # warm near-black (used on light backgrounds). `best_ink` picks
  # whichever gives a higher contrast ratio.
  #
  # For palettes whose primary sits in the mid-luminance range
  # (mustard, peach) neither default passes WCAG AA body-text
  # (4.5:1) — the `Storefronts::Palette` table ships a hand-tuned
  # deep-brown ink for those two cases. This service is for
  # everywhere else: arbitrary hexes, user-picked colors, runtime
  # shading of brand backgrounds. Check `good_enough?` before
  # blindly trusting `best_ink`.
  module Contrast
    INK_LIGHT = "#F8F6F1".freeze  # bone — on dark primaries
    INK_DARK  = "#0E1714".freeze  # warm near-black — on light primaries

    # Minimum ratio for WCAG AA body text (normal-size).
    WCAG_AA_BODY = 4.5
    # Minimum ratio for WCAG AA large text (18pt / 14pt bold).
    WCAG_AA_LARGE = 3.0

    module_function

    # Relative luminance per WCAG 2.1 — floats 0..1.
    def relative_luminance(hex)
      r, g, b = hex.sub("#", "").scan(/../).map { |h| h.to_i(16) / 255.0 }
      linear = [ r, g, b ].map do |c|
        c <= 0.03928 ? c / 12.92 : ((c + 0.055) / 1.055)**2.4
      end
      (0.2126 * linear[0]) + (0.7152 * linear[1]) + (0.0722 * linear[2])
    end

    # WCAG contrast ratio — float 1..21. Order-independent.
    def wcag_ratio(fg, bg)
      lfg = relative_luminance(fg)
      lbg = relative_luminance(bg)
      l1, l2 = [ lfg, lbg ].max, [ lfg, lbg ].min
      (l1 + 0.05) / (l2 + 0.05)
    end

    # Best-of-two text color for a given background hex. Always returns
    # one of the Kitchef ink tokens — if you need a better contrast than
    # either default can offer (mustard-on-cream territory), fall back
    # to a per-palette hand-tuned value.
    def best_ink(background)
      dark  = wcag_ratio(INK_DARK,  background)
      light = wcag_ratio(INK_LIGHT, background)
      dark >= light ? INK_DARK : INK_LIGHT
    end

    # True when the best achievable ink against `background` passes
    # WCAG AA for body text. Use this to decide whether `best_ink` is
    # safe or whether the caller should supply a palette-specific ink.
    def good_enough?(background, threshold: WCAG_AA_BODY)
      ratio = [ wcag_ratio(INK_DARK, background), wcag_ratio(INK_LIGHT, background) ].max
      ratio >= threshold
    end
  end
end

File: app/javascript/lib/contrast.js
// WCAG 2.1 contrast utilities — JS mirror of `Ui::Contrast` in Ruby.
// Used by storefront / account-edit previews where text color needs
// to be picked against a brand background that may change at runtime.
//
//   import { bestInk, wcagRatio } from "lib/contrast"
//   button.style.color = bestInk("#0A5A3C")      // "#F8F6F1"
//   wcagRatio("#F8F6F1", "#0A5A3C")              // ≈ 6.78
//
// Stays in sync with the Ruby constants (INK_LIGHT = bone,
// INK_DARK = warm near-black); changes to the defaults should land
// in both files.

export const INK_LIGHT = "#F8F6F1"  // bone — on dark backgrounds
export const INK_DARK  = "#0E1714"  // warm near-black — on light backgrounds

export const WCAG_AA_BODY  = 4.5
export const WCAG_AA_LARGE = 3.0

// Relative luminance per WCAG 2.1 — number 0..1.
export function relativeLuminance(hex) {
  const parts = hex.replace("#", "").match(/../g).map(h => parseInt(h, 16) / 255)
  const linear = parts.map(c =>
    c <= 0.03928 ? c / 12.92 : Math.pow((c + 0.055) / 1.055, 2.4)
  )
  return 0.2126 * linear[0] + 0.7152 * linear[1] + 0.0722 * linear[2]
}

// WCAG contrast ratio between two hex colors — number 1..21.
export function wcagRatio(fg, bg) {
  const lfg = relativeLuminance(fg)
  const lbg = relativeLuminance(bg)
  const l1 = Math.max(lfg, lbg)
  const l2 = Math.min(lfg, lbg)
  return (l1 + 0.05) / (l2 + 0.05)
}

// Best-of-two ink token for a given background. See `good_enough`
// before trusting the result on mid-luminance primaries (mustard,
// peach) — some palettes need a hand-tuned ink to pass AA.
export function bestInk(background) {
  const dark  = wcagRatio(INK_DARK,  background)
  const light = wcagRatio(INK_LIGHT, background)
  return dark >= light ? INK_DARK : INK_LIGHT
}

export function goodEnough(background, { threshold = WCAG_AA_BODY } = {}) {
  const ratio = Math.max(
    wcagRatio(INK_DARK,  background),
    wcagRatio(INK_LIGHT, background)
  )
  return ratio >= threshold
}
